File: Gateway/Converter/StatValueConverter.cpp
#include "pch.h"
#include "StatValueConverter.h"

#include "StatTypeConverter.h"

StatValue StatValueConverter::toNative(StatValueDto^ dto)
{
	StatValue native;

	if (dto == nullptr) {
		native.Type = StatType::None;
		native.Value = 0.0;
		return native;
	}

	native.Type = StatTypeConverter::toNative(dto->Type);
	native.Value = dto->Value;

	return native;
}

StatValueDto^ StatValueConverter::toDto(const StatValue& native)
{
	StatValueDto^ dto = gcnew StatValueDto();

	dto->Type = StatTypeConverter::toDto(native.Type);
	dto->Value = native.Value;

	return dto;
}

File: Gui/Converter/StatValueConverter.cs
using Gui.Model;
using SaveData.Model;

namespace Gui.Converter
{
	public static class StatValueConverter
	{
		public static StatValueSaveData ToSaveData(StatValueDto dto)
		{
			if (dto == null)
			{
				return null;
			}

			return new StatValueSaveData
			{
				Type = StatTypeConverter.ToText(dto.Type),
				Value = dto.Value
			};
		}

		public static StatValueSaveData ToSaveData(StatValueInputModel inputModel)
		{
			if (inputModel == null)
			{
				return null;
			}

			if (string.IsNullOrWhiteSpace(inputModel.Type) && string.IsNullOrWhiteSpace(inputModel.ValueText))
			{
				return new StatValueSaveData
				{
					Type = inputModel.Type,
					Value = 0
				};
			}

			return new StatValueSaveData
			{
				Type = inputModel.Type,
				Value = inputModel.Value
			};
		}

		public static StatValueDto ToDto(StatValueInputModel inputModel)
		{
			if (inputModel == null)
			{
				return null;
			}

			if (string.IsNullOrWhiteSpace(inputModel.Type) && string.IsNullOrWhiteSpace(inputModel.ValueText))
			{
				return new StatValueDto
				{
					Type = StatTypeDto.None,
					Value = 0
				};
			}

			return new StatValueDto
			{
				Type = StatTypeConverter.ToDto(inputModel.Type),
				Value = inputModel.Value
			};
		}

		public static StatValueInputModel ToInputModel(StatValueSaveData saveData)
		{
			if (saveData == null)
			{
				return null;
			}

			return new StatValueInputModel
			{
				Type = saveData.Type,
				ValueText = saveData.Value.ToString()
			};
		}
	}
}
